File: java17/src/main/java/org/chelck/dp/BlockingTable.java
package org.chelck.dp;

import java.util.Objects;
import java.util.Optional;

public class BlockingTable {
    private final Object lock;
    private final Table table;

    public BlockingTable(Table table) {
        this.lock = new Object();
        this.table = Objects.requireNonNull(table);
    }

    public int getLeftFork(int philosopher) {
        synchronized (lock) {
            while (true) {
                Optional<Integer> f = this.table.getLeftFork(philosopher);
                if (f.isPresent()) {
                    return f.get();
                }
                else {
                    try {
                        lock.wait();
                    }
                    catch (InterruptedException e) {

                    }
                }
            }
        }
    }

    public void returnLeftFork(int philosopher) {
        synchronized (lock) {
            this.table.returnLeftFork(philosopher);
            lock.notifyAll();
        }
    }

    public int getRightFork(int philosopher) {
        synchronized (lock) {
            while (true) {
                Optional<Integer> f = this.table.getRightFork(philosopher);
                if (f.isPresent()) {
                    return f.get();
                }
                else {
                    try {
                        lock.wait();
                    }
                    catch (InterruptedException e) {

                    }
                }
            }
        }
    }

    public void returnRightFork(int philosopher) {
        synchronized (lock) {
            this.table.returnRightFork(philosopher);
            lock.notifyAll();
        }
    }


}

File: scala3_mutex/src/main/scala/org/chelck/dp/BlockingTable.scala
package org.chelck.dp

class BlockingTable(val table: Table) {
  val lock = new Object

  def getLeftFork(philosopher: Int): Int = {
    lock.synchronized {
      var f: Option[Int] = table.getLeftFork(philosopher)

      while (f == None) {
        f = table.getLeftFork(philosopher)
        if (f.isEmpty) lock.wait()
      }
      f.get
    }
  }

  def returnLeftFork(philosopher: Int): Unit = {
    lock synchronized {
      table.returnLeftFork(philosopher)
      lock.notifyAll()
    }

  }

  def getRightFork(philosopher: Int): Int = {
    lock.synchronized {
      var f: Option[Int] = table.getRightFork(philosopher)

      while (f == None) {
        f = table.getRightFork(philosopher)
        if (f.isEmpty) lock.wait()
      }
      f.get
    }
  }

  def returnRightFork(philosopher: Int): Unit = {
    lock synchronized {
      table.returnRightFork(philosopher)
      lock.notifyAll()
    }
  }
}
